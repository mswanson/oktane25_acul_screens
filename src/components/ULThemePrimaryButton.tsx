import { Button, type ButtonProps } from "@/components/ui/button";
import { cn } from "@/lib/utils";

export type ULThemePrimaryButtonProps = ButtonProps;

export function ULThemePrimaryButton({
  variant = "primary",
  size = "default",
  className,
  ...props
}: ULThemePrimaryButtonProps) {
  // Variant-specific theme overrides for colors and states
  const variantThemeOverrides = {
    primary: cn(
      "p-6 border-0",
      "cursor-pointer",
      "theme-universal:bg-primary-button",
      "theme-universal:text-(--ul-theme-color-primary-button-label)", //text-color
      "theme-universal:hover:shadow-[var(--button-hover-shadow)]",
      "theme-universal:focus:outline-none theme-universal:focus:ring-4 theme-universal:focus:ring-base-focus/15",
      "theme-universal:disabled:bg-primary-button/70",
      "theme-universal:disabled:border-primary-button/70",
      "theme-universal:disabled:cursor-not-allowed"
    ),
    secondary: "", // Add secondary overrides if needed
    destructive: "", // Add destructive overrides if needed
    outline: "", // Add outline overrides if needed
    ghost: "", // Add ghost overrides if needed
    link: "", // Add link overrides if needed
  };

  // Size-specific theme overrides for border radius and typography
  const sizeThemeOverrides = {
    default: cn(
      "theme-universal:rounded-button",
      "theme-universal:font-button", //font-weight
      "theme-universal:text-(length:--ul-theme-font-buttons-text-size)" //font-size
    ),
    xs: cn(
      "theme-universal:rounded-button",
      "theme-universal:font-button",
      "theme-universal:text-(length:--ul-theme-font-buttons-text-size)"
    ),
    sm: cn(
      "theme-universal:rounded-button",
      "theme-universal:font-button",
      "theme-universal:text-(length:--ul-theme-font-buttons-text-size)"
    ),
    lg: cn(
      "theme-universal:rounded-button",
      "theme-universal:font-button",
      "theme-universal:text-(length:--ul-theme-font-buttons-text-size)"
    ),
    icon: cn("theme-universal:rounded-button"),
  };

  // Combine all theme classes with proper type safety
  const themeClasses = cn(
    variant && variantThemeOverrides[variant],
    size && sizeThemeOverrides[size]
  );

  return (
    <Button
      variant={variant}
      size={size}
      className={cn(className, themeClasses)}
      {...props}
    />
  );
}
