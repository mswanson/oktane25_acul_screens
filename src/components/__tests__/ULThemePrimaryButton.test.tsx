import { render, screen } from "@testing-library/react";

import { ULThemePrimaryButton } from "../ULThemePrimaryButton";

describe("ULThemePrimaryButton", () => {
  // Snapshot Tests
  it("matches snapshot with default props", () => {
    const { container } = render(
      <ULThemePrimaryButton>Default Button</ULThemePrimaryButton>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("matches snapshot with disabled state", () => {
    const { container } = render(
      <ULThemePrimaryButton disabled>Disabled Button</ULThemePrimaryButton>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("matches snapshot with custom className", () => {
    const { container } = render(
      <ULThemePrimaryButton className="w-full custom-btn">
        Full Width Button
      </ULThemePrimaryButton>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("matches snapshot with different sizes", () => {
    const { container: smallContainer } = render(
      <ULThemePrimaryButton size="sm">Small Button</ULThemePrimaryButton>
    );
    const { container: largeContainer } = render(
      <ULThemePrimaryButton size="lg">Large Button</ULThemePrimaryButton>
    );

    expect(smallContainer.firstChild).toMatchSnapshot("small-button");
    expect(largeContainer.firstChild).toMatchSnapshot("large-button");
  });

  // Functional Tests
  it("renders with text content", () => {
    render(<ULThemePrimaryButton>Click me</ULThemePrimaryButton>);
    expect(
      screen.getByRole("button", { name: "Click me" })
    ).toBeInTheDocument();
  });

  it("forwards props to UDS Button", () => {
    render(
      <ULThemePrimaryButton disabled data-testid="test-button">
        Disabled Button
      </ULThemePrimaryButton>
    );
    const button = screen.getByTestId("test-button");
    expect(button).toBeDisabled();
  });

  it("applies theme classes correctly", () => {
    render(
      <ULThemePrimaryButton data-testid="themed-button">
        Themed
      </ULThemePrimaryButton>
    );
    const button = screen.getByTestId("themed-button");
    expect(button.className).toContain("theme-universal:bg-primary-button");
  });

  it("supports custom className", () => {
    render(
      <ULThemePrimaryButton
        className="custom-class"
        data-testid="custom-button"
      >
        Custom
      </ULThemePrimaryButton>
    );
    const button = screen.getByTestId("custom-button");
    expect(button).toHaveClass("custom-class");
  });

  it("has proper accessibility attributes", () => {
    render(<ULThemePrimaryButton>Accessible Button</ULThemePrimaryButton>);
    const button = screen.getByRole("button");
    expect(button).toBeInTheDocument();
    expect(button.tagName.toLowerCase()).toBe("button");
  });

  it("applies Auth0 theme utility classes", () => {
    render(
      <ULThemePrimaryButton data-testid="auth0-button">
        Auth0 Themed
      </ULThemePrimaryButton>
    );
    const button = screen.getByTestId("auth0-button");

    // Check for theme-universal variant classes
    expect(button.className).toContain("theme-universal:bg-primary-button");
    expect(button.className).toContain("theme-universal:font-button");
    expect(button.className).toContain("theme-universal:rounded-button");
  });
});
