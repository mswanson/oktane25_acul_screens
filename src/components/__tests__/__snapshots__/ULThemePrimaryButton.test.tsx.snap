// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ULThemePrimaryButton matches snapshot with custom className 1`] = `
<button
  class="focus-visible:ring-ring aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive theme-default:active:scale-[0.99] relative box-border inline-flex shrink-0 items-center justify-center gap-2 overflow-hidden text-sm font-medium whitespace-nowrap transition-all duration-150 ease-in-out outline-none focus-visible:ring-4 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4 shadow-button-resting hover:shadow-button-hover hover:border-primary/50 border-primary bg-primary text-primary-foreground hover:bg-primary/90 theme-default:before:from-primary-foreground/0 theme-default:before:to-primary-foreground/15 theme-default:before:absolute theme-default:before:top-0 theme-default:before:left-0 theme-default:before:block theme-default:before:h-full theme-default:before:w-full theme-default:before:bg-gradient-to-t theme-default:before:content-[''] h-10 rounded-2xl has-[>svg]:px-3 w-full custom-btn p-6 border-0 cursor-pointer theme-universal:bg-primary-button theme-universal:text-(--ul-theme-color-primary-button-label) theme-universal:hover:shadow-[var(--button-hover-shadow)] theme-universal:focus:outline-none theme-universal:focus:ring-4 theme-universal:focus:ring-base-focus/15 theme-universal:disabled:bg-primary-button/70 theme-universal:disabled:border-primary-button/70 theme-universal:disabled:cursor-not-allowed theme-universal:rounded-button theme-universal:font-button theme-universal:text-(length:--ul-theme-font-buttons-text-size)"
>
  Full Width Button
</button>
`;

exports[`ULThemePrimaryButton matches snapshot with default props 1`] = `
<button
  class="focus-visible:ring-ring aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive theme-default:active:scale-[0.99] relative box-border inline-flex shrink-0 items-center justify-center gap-2 overflow-hidden text-sm font-medium whitespace-nowrap transition-all duration-150 ease-in-out outline-none focus-visible:ring-4 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4 shadow-button-resting hover:shadow-button-hover hover:border-primary/50 border-primary bg-primary text-primary-foreground hover:bg-primary/90 theme-default:before:from-primary-foreground/0 theme-default:before:to-primary-foreground/15 theme-default:before:absolute theme-default:before:top-0 theme-default:before:left-0 theme-default:before:block theme-default:before:h-full theme-default:before:w-full theme-default:before:bg-gradient-to-t theme-default:before:content-[''] h-10 rounded-2xl has-[>svg]:px-3 p-6 border-0 cursor-pointer theme-universal:bg-primary-button theme-universal:text-(--ul-theme-color-primary-button-label) theme-universal:hover:shadow-[var(--button-hover-shadow)] theme-universal:focus:outline-none theme-universal:focus:ring-4 theme-universal:focus:ring-base-focus/15 theme-universal:disabled:bg-primary-button/70 theme-universal:disabled:border-primary-button/70 theme-universal:disabled:cursor-not-allowed theme-universal:rounded-button theme-universal:font-button theme-universal:text-(length:--ul-theme-font-buttons-text-size)"
>
  Default Button
</button>
`;

exports[`ULThemePrimaryButton matches snapshot with different sizes: large-button 1`] = `
<button
  class="focus-visible:ring-ring aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive theme-default:active:scale-[0.99] relative box-border inline-flex shrink-0 items-center justify-center gap-2 overflow-hidden font-medium whitespace-nowrap transition-all duration-150 ease-in-out outline-none focus-visible:ring-4 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4 shadow-button-resting hover:shadow-button-hover hover:border-primary/50 border-primary bg-primary text-primary-foreground hover:bg-primary/90 theme-default:before:from-primary-foreground/0 theme-default:before:to-primary-foreground/15 theme-default:before:absolute theme-default:before:top-0 theme-default:before:left-0 theme-default:before:block theme-default:before:h-full theme-default:before:w-full theme-default:before:bg-gradient-to-t theme-default:before:content-[''] h-12 rounded-3xl text-base has-[>svg]:px-4 p-6 border-0 cursor-pointer theme-universal:bg-primary-button theme-universal:text-(--ul-theme-color-primary-button-label) theme-universal:hover:shadow-[var(--button-hover-shadow)] theme-universal:focus:outline-none theme-universal:focus:ring-4 theme-universal:focus:ring-base-focus/15 theme-universal:disabled:bg-primary-button/70 theme-universal:disabled:border-primary-button/70 theme-universal:disabled:cursor-not-allowed theme-universal:rounded-button theme-universal:font-button theme-universal:text-(length:--ul-theme-font-buttons-text-size)"
>
  Large Button
</button>
`;

exports[`ULThemePrimaryButton matches snapshot with different sizes: small-button 1`] = `
<button
  class="focus-visible:ring-ring aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive theme-default:active:scale-[0.99] relative box-border inline-flex shrink-0 items-center justify-center overflow-hidden font-medium whitespace-nowrap transition-all duration-150 ease-in-out outline-none focus-visible:ring-4 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4 shadow-button-resting hover:shadow-button-hover hover:border-primary/50 border-primary bg-primary text-primary-foreground hover:bg-primary/90 theme-default:before:from-primary-foreground/0 theme-default:before:to-primary-foreground/15 theme-default:before:absolute theme-default:before:top-0 theme-default:before:left-0 theme-default:before:block theme-default:before:h-full theme-default:before:w-full theme-default:before:bg-gradient-to-t theme-default:before:content-[''] h-8 gap-1.5 rounded-xl text-xs has-[>svg]:px-2.5 p-6 border-0 cursor-pointer theme-universal:bg-primary-button theme-universal:text-(--ul-theme-color-primary-button-label) theme-universal:hover:shadow-[var(--button-hover-shadow)] theme-universal:focus:outline-none theme-universal:focus:ring-4 theme-universal:focus:ring-base-focus/15 theme-universal:disabled:bg-primary-button/70 theme-universal:disabled:border-primary-button/70 theme-universal:disabled:cursor-not-allowed theme-universal:rounded-button theme-universal:font-button theme-universal:text-(length:--ul-theme-font-buttons-text-size)"
>
  Small Button
</button>
`;

exports[`ULThemePrimaryButton matches snapshot with disabled state 1`] = `
<button
  class="focus-visible:ring-ring aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive theme-default:active:scale-[0.99] relative box-border inline-flex shrink-0 items-center justify-center gap-2 overflow-hidden text-sm font-medium whitespace-nowrap transition-all duration-150 ease-in-out outline-none focus-visible:ring-4 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4 shadow-button-resting hover:shadow-button-hover hover:border-primary/50 border-primary bg-primary text-primary-foreground hover:bg-primary/90 theme-default:before:from-primary-foreground/0 theme-default:before:to-primary-foreground/15 theme-default:before:absolute theme-default:before:top-0 theme-default:before:left-0 theme-default:before:block theme-default:before:h-full theme-default:before:w-full theme-default:before:bg-gradient-to-t theme-default:before:content-[''] h-10 rounded-2xl has-[>svg]:px-3 p-6 border-0 cursor-pointer theme-universal:bg-primary-button theme-universal:text-(--ul-theme-color-primary-button-label) theme-universal:hover:shadow-[var(--button-hover-shadow)] theme-universal:focus:outline-none theme-universal:focus:ring-4 theme-universal:focus:ring-base-focus/15 theme-universal:disabled:bg-primary-button/70 theme-universal:disabled:border-primary-button/70 theme-universal:disabled:cursor-not-allowed theme-universal:rounded-button theme-universal:font-button theme-universal:text-(length:--ul-theme-font-buttons-text-size)"
  disabled=""
>
  Disabled Button
</button>
`;
